# EDA

In [1]:
import matplotlib.pyplot as plt
import seaborn as sn
import pandas as pd
import numpy as np

In [2]:
df1 = pd.read_csv(r"D:\Unemployment in India.csv")

Data Frame 1(df1) represent Unemployment in India

In [3]:
df2 = pd.read_csv(r"D:\Unemployment_Rate_upto_11_2020.csv")

Data Frame 2 (df2) represent Unemployment Rate in 2020 table

In [5]:
df1.head()

Region         Date  Frequency   Estimated Unemployment Rate (%)  \
0  Andhra Pradesh   31-05-2019    Monthly                              3.65   
1  Andhra Pradesh   30-06-2019    Monthly                              3.05   
2  Andhra Pradesh   31-07-2019    Monthly                              3.75   
3  Andhra Pradesh   31-08-2019    Monthly                              3.32   
4  Andhra Pradesh   30-09-2019    Monthly                              5.17   

    Estimated Employed   Estimated Labour Participation Rate (%)   Area  
0           11999139.0                                     43.24  Rural  
1           11755881.0                                     42.05  Rural  
2           12086707.0                                     43.50  Rural  
3           12285693.0                                     43.97  Rural  
4           12256762.0                                     44.68  Rural

In [6]:
df2.head()

Region         Date  Frequency   Estimated Unemployment Rate (%)  \
0  Andhra Pradesh   31-01-2020          M                              5.48   
1  Andhra Pradesh   29-02-2020          M                              5.83   
2  Andhra Pradesh   31-03-2020          M                              5.79   
3  Andhra Pradesh   30-04-2020          M                             20.51   
4  Andhra Pradesh   31-05-2020          M                             17.43   

    Estimated Employed   Estimated Labour Participation Rate (%) Region.1  \
0             16635535                                     41.02    South   
1             16545652                                     40.90    South   
2             15881197                                     39.18    South   
3             11336911                                     33.10    South   
4             12988845                                     36.46    South   

   longitude  latitude  
0    15.9129     79.74  
1    15.9129     79.74  
2    15.9129     79.74  
3    15.9129     79.74  
4    15.9129     79.74

# meaningful questions about the dataset before analysis

In [ ]:
Unemployment Analysis in India (2019–2020)

## Goals:
- Analyze trends in unemployment across states and time.
- Compare unemployment between urban and rural areas.
- Detect anomalies during the COVID-19 period.

# Date Structure

In [8]:
df1 = pd.read_csv(r"D:\Unemployment in India.csv")
df1.columns

Index(['Region', ' Date', ' Frequency', ' Estimated Unemployment Rate (%)',
       ' Estimated Employed', ' Estimated Labour Participation Rate (%)',
       'Area'],
      dtype='object')

As we can see there are extra spaces in the columns

In [9]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 7 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   Region                                    740 non-null    object 
 1    Date                                     740 non-null    object 
 2    Frequency                                740 non-null    object 
 3    Estimated Unemployment Rate (%)          740 non-null    float64
 4    Estimated Employed                       740 non-null    float64
 5    Estimated Labour Participation Rate (%)  740 non-null    float64
 6   Area                                      740 non-null    object 
dtypes: float64(3), object(4)
memory usage: 42.1+ KB


There are some cleaning requirements in 'Date column' and 'Estimated Employed' column

In [10]:
df2 = pd.read_csv(r"D:\Unemployment_Rate_upto_11_2020.csv")
print(df2.columns)

Index(['Region', ' Date', ' Frequency', ' Estimated Unemployment Rate (%)',
       ' Estimated Employed', ' Estimated Labour Participation Rate (%)',
       'Region.1', 'longitude', 'latitude'],
      dtype='object')


As we can see there are extra spaces in the columns

In [11]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 267 entries, 0 to 266
Data columns (total 9 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   Region                                    267 non-null    object 
 1    Date                                     267 non-null    object 
 2    Frequency                                267 non-null    object 
 3    Estimated Unemployment Rate (%)          267 non-null    float64
 4    Estimated Employed                       267 non-null    int64  
 5    Estimated Labour Participation Rate (%)  267 non-null    float64
 6   Region.1                                  267 non-null    object 
 7   longitude                                 267 non-null    float64
 8   latitude                                  267 non-null    float64
dtypes: float64(4), int64(1), object(4)
memory usage: 18.9+ KB


There are some cleaning requirements in 'Date' column

# Cleaning Datas and Extra Column spaces Both Table Unemployment in India and Unemployment Rate in Year 2020

In [12]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 7 columns):
 #   Column                                    Non-Null Count  Dtype  
---  ------                                    --------------  -----  
 0   Region                                    740 non-null    object 
 1    Date                                     740 non-null    object 
 2    Frequency                                740 non-null    object 
 3    Estimated Unemployment Rate (%)          740 non-null    float64
 4    Estimated Employed                       740 non-null    float64
 5    Estimated Labour Participation Rate (%)  740 non-null    float64
 6   Area                                      740 non-null    object 
dtypes: float64(3), object(4)
memory usage: 42.1+ KB


In [13]:
df1.describe

<bound method NDFrame.describe of              Region         Date  Frequency   Estimated Unemployment Rate (%)  \
0    Andhra Pradesh   31-05-2019    Monthly                              3.65   
1    Andhra Pradesh   30-06-2019    Monthly                              3.05   
2    Andhra Pradesh   31-07-2019    Monthly                              3.75   
3    Andhra Pradesh   31-08-2019    Monthly                              3.32   
4    Andhra Pradesh   30-09-2019    Monthly                              5.17   
..              ...          ...        ...                               ...   
763             NaN          NaN        NaN                               NaN   
764             NaN          NaN        NaN                               NaN   
765             NaN          NaN        NaN                               NaN   
766             NaN          NaN        NaN                               NaN   
767             NaN          NaN        NaN                               N

In [18]:
#1 Changing Date type from object to Date and Estimated Employed from float to int in Unemployment in India table
#2 Adjustment columns extra spaces
print(df1.columns.tolist())
df1.columns = df1.columns.str.strip()
df1['Date'] = pd.to_datetime(df1['Date'],dayfirst = True)
df1.info()

['Region', 'Date', 'Frequency', 'Estimated Unemployment Rate (%)', 'Estimated Employed', 'Estimated Labour Participation Rate (%)', 'Area']
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 7 columns):
 #   Column                                   Non-Null Count  Dtype         
---  ------                                   --------------  -----         
 0   Region                                   740 non-null    object        
 1   Date                                     740 non-null    datetime64[ns]
 2   Frequency                                740 non-null    object        
 3   Estimated Unemployment Rate (%)          740 non-null    float64       
 4   Estimated Employed                       740 non-null    float64       
 5   Estimated Labour Participation Rate (%)  740 non-null    float64       
 6   Area                                     740 non-null    object        
dtypes: datetime64[ns](1), float64(3), object(3)
memory usage: 

In [21]:
#1 Changing Date type from object to Date For Unemployment rate in year 2020
#2 Adjustment columns extra spaces
print(df2.columns.tolist())
df2.columns = df2.columns.str.strip()
df2['Date'] = pd.to_datetime(df2['Date'],dayfirst = True)
df2.info()

['Region', ' Date', ' Frequency', ' Estimated Unemployment Rate (%)', ' Estimated Employed', ' Estimated Labour Participation Rate (%)', 'Region.1', 'longitude', 'latitude']
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 267 entries, 0 to 266
Data columns (total 9 columns):
 #   Column                                   Non-Null Count  Dtype         
---  ------                                   --------------  -----         
 0   Region                                   267 non-null    object        
 1   Date                                     267 non-null    datetime64[ns]
 2   Frequency                                267 non-null    object        
 3   Estimated Unemployment Rate (%)          267 non-null    float64       
 4   Estimated Employed                       267 non-null    int64         
 5   Estimated Labour Participation Rate (%)  267 non-null    float64       
 6   Region.1                                 267 non-null    object        
 7   longitude               

In [22]:
#3 Checking Null Values and remove it in in Unemployment in India table
df1.isnull().sum()
print("\nRows and Columns:") 
print("Before with Null Values" ,df1.shape) 
df1 = df1.dropna()
print("After Remove Null Values", df1.shape)


Rows and Columns:
Before with Null Values (768, 7)
After Remove Null Values (740, 7)


In [23]:
#4 Checking Null Values and remove it in in Unemployment  Rate in Year 2020
df2.isnull().sum()
print("\nRows and Columns:") 
print("Before with Null Values" ,df2.shape) 
df2 = df2.dropna()
print("After Remove Null Values", df2.shape)


Rows and Columns:
Before with Null Values (267, 9)
After Remove Null Values (267, 9)


# Identify trends, patterns and anomalies within the data

In [24]:
print("\nNumber of Uniques Regions:")
print(df1['Region'].nunique())
print("\nName of Unique Regions:")
print(df1['Region'].unique())
print("\nArea Distribution:")
print(df1['Area'].unique())
print(df1['Area'].value_counts())


Number of Uniques Regions:
28

Name of Unique Regions:
['Andhra Pradesh' 'Assam' 'Bihar' 'Chhattisgarh' 'Delhi' 'Goa' 'Gujarat'
 'Haryana' 'Himachal Pradesh' 'Jammu & Kashmir' 'Jharkhand' 'Karnataka'
 'Kerala' 'Madhya Pradesh' 'Maharashtra' 'Meghalaya' 'Odisha' 'Puducherry'
 'Punjab' 'Rajasthan' 'Sikkim' 'Tamil Nadu' 'Telangana' 'Tripura'
 'Uttar Pradesh' 'Uttarakhand' 'West Bengal' 'Chandigarh']

Area Distribution:
['Rural' 'Urban']
Area
Urban    381
Rural    359
Name: count, dtype: int64


In [25]:
# Labour Fource are more in rural areas then urban areas
labour_force = df1.groupby('Area')['Estimated Employed'].sum().sort_values(ascending = False)
print("n\Total Employeed:",  labour_force)
labour_force2 = df1.groupby('Area')['Estimated Labour Participation Rate (%)'].mean().sort_values(ascending = False)
print("\nAvg labour participation rate:",labour_force2)

n\Total Employeed: Area
Rural    3.659234e+09
Urban    1.672066e+09
Name: Estimated Employed, dtype: float64

Avg labour participation rate: Area
Rural    44.464819
Urban    40.901365
Name: Estimated Labour Participation Rate (%), dtype: float64


<>:3: SyntaxWarning: invalid escape sequence '\T'
<>:3: SyntaxWarning: invalid escape sequence '\T'
C:\Users\rajne\AppData\Local\Temp\ipykernel_18880\768626678.py:3: SyntaxWarning: invalid escape sequence '\T'
  print("n\Total Employeed:",  labour_force)


In [26]:
Region_Wise_Avg_Unemployment = df1.groupby('Region')['Estimated Unemployment Rate (%)'].mean().sort_values(ascending = False).round(2)
print("\nRegion wise Avg Unemployment Rate:",  Region_Wise_Avg_Unemployment)


Region wise Avg Unemployment Rate: Region
Tripura             28.35
Haryana             26.28
Jharkhand           20.58
Bihar               18.92
Himachal Pradesh    18.54
Delhi               16.50
Jammu & Kashmir     16.19
Chandigarh          15.99
Rajasthan           14.06
Uttar Pradesh       12.55
Punjab              12.03
Puducherry          10.22
Kerala              10.12
Tamil Nadu           9.28
Goa                  9.27
Chhattisgarh         9.24
West Bengal          8.12
Telangana            7.74
Maharashtra          7.56
Andhra Pradesh       7.48
Madhya Pradesh       7.41
Sikkim               7.25
Karnataka            6.68
Gujarat              6.66
Uttarakhand          6.58
Assam                6.43
Odisha               5.66
Meghalaya            4.80
Name: Estimated Unemployment Rate (%), dtype: float64


AS we can see Higest Unemployment rate in 'Tripura', And Lowest Unemployment rate in 'Meghalaya'

In [27]:
# Checking anomalies
anomalies = df1[df1['Estimated Unemployment Rate (%)']>50].reset_index()
print(anomalies)

   index      Region       Date Frequency  Estimated Unemployment Rate (%)  \
0    145   Jharkhand 2020-05-31   Monthly                            55.10   
1    242  Puducherry 2020-04-30   Monthly                            74.51   
2    287  Tamil Nadu 2020-04-30   Monthly                            53.19   
3    412       Bihar 2020-04-30   Monthly                            58.77   
4    530   Jharkhand 2020-04-30   Monthly                            61.48   
5    531   Jharkhand 2020-05-31   Monthly                            70.17   
6    627  Puducherry 2020-04-30   Monthly                            76.74   
7    628  Puducherry 2020-05-31   Monthly                            75.00   

   Estimated Employed  Estimated Labour Participation Rate (%)   Area  
0           3315038.0                                    37.03  Rural  
1             49420.0                                    46.79  Rural  
2           5086200.0                                    33.05  Rural  
3        

As we can see there are vary high unemployment rate in 'Puducherry'which is 76.74 % Region

# Testing Hypotheses

In [28]:
#1. Unemployment are more in urban areas compared to rural areas
Urban_and_Rural_unemployment_avg = df1.groupby('Area')['Estimated Unemployment Rate (%)'].mean()
print(Urban_and_Rural_unemployment_avg)

Area
Rural    10.324791
Urban    13.166614
Name: Estimated Unemployment Rate (%), dtype: float64


Null Hypothesis Ho: Avg rural unemployment rate are > then urban unemplyment rate Alternat Hypothesis H1 : Avg urban unemployment rate are > then rural unemployment rate AS we can see Urban Unemploment avg rate is > then rural avg, So we can say that urban areas faces more unemployment then ruela areas, As result we fail to reject the alternat hypothesis H1 Insight: Urban unemployment rates are on average 3% higher than rural areas.

In [29]:
#2. COVID 19 lockdown impact on employment in (April 2020)
print("\nAvg Unemployment rate befor lockdown year (2019) was:", df1['Estimated Unemployment Rate (%)'].mean().round(2))
April_2020 = df2[df2['Date']=='2020-04-30']
print("\nUnemployment from April 2020:\n" , April_2020[['Region','Date','Estimated Unemployment Rate (%)']].sort_values(by = 'Estimated Unemployment Rate (%)',ascending = False ))


Avg Unemployment rate befor lockdown year (2019) was: 11.79

Unemployment from April 2020:
                Region       Date  Estimated Unemployment Rate (%)
172        Puducherry 2020-04-30                            75.85
210        Tamil Nadu 2020-04-30                            49.83
102         Jharkhand 2020-04-30                            47.09
23              Bihar 2020-04-30                            46.64
73            Haryana 2020-04-30                            43.22
230           Tripura 2020-04-30                            41.23
112         Karnataka 2020-04-30                            29.84
162            Odisha 2020-04-30                            23.76
240     Uttar Pradesh 2020-04-30                            21.54
142       Maharashtra 2020-04-30                            20.90
3      Andhra Pradesh 2020-04-30                            20.51
63            Gujarat 2020-04-30                            18.71
192         Rajasthan 2020-04-30                 


Null Hypothesis Ho: There are not any impact in unemployment rate due to COVID 19 Starting month (2020-04-30) Alternat Hypothesis H1: There are negative impact in unemployment rate dut to COVID 19 lockdown Started month (2020-04-30) As we can see when COVID 19 lockdown Started April Month 2020, 'Puducherry' much more affect compare to other states , As we can see the number we are fail to reject Alternat hupothesis means - There are some negative inpact in Unemployment rate due to Lockdown Insight: Puducherry had the highest spike (75%) in unemployment in April 2020, likely due to strict lockdown effects.

In [33]:
# COVID 19 affect unemployment rate
df2['Months'] = df2['Date'].dt.month
df2.groupby(['Months'])['Estimated Unemployment Rate (%)'].mean()
print("\nCOVID started year 2020 Monthly unemployment avg rate:", df2.groupby(['Months'])['Estimated Unemployment Rate (%)'].mean().round(2).sort_values (ascending =False ))


COVID started year 2020 Monthly unemployment avg rate: Months
5     23.24
4     22.24
6     10.91
3     10.78
8     10.31
7      9.83
2      9.27
1      9.20
9      8.71
10     8.03
Name: Estimated Unemployment Rate (%), dtype: float64


As we can see in COVID year 2020 Started lockdown month April(4) to June(6) there are pick unemployment rate in India

In [31]:
# Unemployment trend Over time
df1['Months'] = df1['Date'].dt.month
df1.groupby('Months')['Estimated Unemployment Rate (%)'].mean().round(2)

C:\Users\rajne\AppData\Local\Temp\ipykernel_18880\1262575362.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['Months'] = df1['Date'].dt.month


Months
1      9.95
2      9.96
3     10.70
4     23.64
5     16.65
6     10.55
7      9.03
8      9.64
9      9.05
10     9.90
11     9.87
12     9.50
Name: Estimated Unemployment Rate (%), dtype: float64

This Picture shows that unemployment rate gradually increased between 3 to 6 in 2019